### ISSUE TITLE
Emulator won't start functions even with minimal repo

### ISSUE THREAD

**Issue by jelling** [firebase/firebase-tools#6463]
### [REQUIRED] Environment info

**firebase-tools:** 12.7.0
**node version**: 18.18.0

**Platform:** MacOS 12.6.7 (Monterey)

### [REQUIRED] Test case

These are the only files in the project. I cut it down to this from a full project that had been working for months prior. 

#### index.ts

```
import * as functions from "firebase-functions";
import { CallableContext } from "firebase-functions/v1/https";

const foo = functions.https.onCall(
  async (data: any, context: CallableContext) => {
    return true;
  }
);

export { foo };
```

#### tsconfig.json

```
{
  "compilerOptions": {
    "allowJs": true,
    "baseUrl": "src",
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "noImplicitReturns": true,
    "noUnusedLocals": false,
    "outDir": "lib",
    "resolveJsonModule": true,
    "sourceMap": true,
    "strict": true,
    "target": "es2017"
  },
  "compileOnSave": true,
  "include": ["src"]
}
```

#### package.json

```
{
  "name": "functions",
  "scripts": {
    "lint": "eslint --ext .js,.ts .",
    "build": "tsc && tsc-alias",
    "build:watch": "tsc && (concurrently \"tsc -w\" \"tsc-alias -w\")",
    "serve": "npm run build && firebase emulators:start --only functions",
    "shell": "npm run build && firebase functions:shell",
    "start": "npm run shell",
    "deploy:functions": "firebase deploy --only functions --debug",
    "logs": "firebase functions:log",
    "test": "jest"
  },
  "engines": {
    "node": "18"
  },
  "main": "lib/index.js",
  "dependencies": {
    "firebase-admin": "^11.5.0",
    "firebase-functions": "^4.4.1"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "concurrently": "^7.6.0",
    "eslint": "^8.9.0",
    "eslint-config-google": "^0.14.0",
    "eslint-plugin-import": "^2.25.4",
    "firebase-functions-test": "^3.1.0",
    "jest": "^29.5.0",
    "jest-html-reporters": "^3.1.4",
    "tsc-alias": "^1.8.3",
    "typescript": "^4.9.0",
    "url": "^0.11.0",
    "util": "^0.12.5"
  },
  "private": true
}
```

### [REQUIRED] Steps to reproduce

- `npm run build` this works without an error
- `npm run serve` build works but then get this error:

As you can see, there is almost nothing in this proiejct. 

### [REQUIRED] Expected behavior

Functions start in emulator

### [REQUIRED] Actual behavior

```
[2023-10-20T09:03:25.646Z] Got response code 400; body Failed to generate manifest from function source: SyntaxError: Cannot use import statement outside a module
shutdown requested via /__/quitquitquit

⬢  functions: Failed to load function definition from source: FirebaseError: Functions codebase could not be analyzed successfully. It may have a syntax or runtime error {"metadata":{"emulator":{"name":"functions"},"message":"Failed to load function definition from source: FirebaseError: Functions codebase could not be analyzed successfully. It may have a syntax or runtime error"}}
```

**Comment by jelling** [comment]
Updated with more info.

**Comment by jelling** [comment]
I created this bug even though I find the fix for my case because the lack of accurate error reporting from the emulator cost me tons of time here. And no doubt, other people as well.

The fix in my case was to change the `tsconfig.json` from 

```
    "module": "esnext",
    "moduleResolution": "node",
```

to

```
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
```    

Is there a technical reason why the emulator couldn't advise me of that?

**Comment by jelling** [comment]
Here's another issue that 1)gives the same error, but 2) the build works fine.

Causes error:
```
import { audio2TreatmentHandler } from "exampleDir/exampleHandler";
```

Works: 

```
import { audio2TreatmentHandler } from "./exampleDir/exampleHandler";
```

Again, I had to debug this via trial and error. And this is particularly frustrating as I don't have these sort of issues with other build pipelines / runtimes. I get that the emulator / cloud function runtime is closed source, but without better error reporting that becomes a real liability.

**Comment by Berlioz** [comment]
Neither the emulator nor the Functions runtime are closed source. The emulator's implementation is in this repository and the Functions runtime repository is at https://github.com/firebase/firebase-functions.

There are limits to how fancy the error handling can be in the emulator for unrunnable code, given that Functions' runtime basically relies on loading the code and inspecting its shape in a first pass. However, running the `emulators:start` command with `--debug` or inspecting the `*-debug.log` files generated would have shown you the raw syntax errors being generated.

**Comment by jelling** [comment]
>  However, running the emulators:start command with --debug or inspecting the *-debug.log files generated would have shown you the raw syntax errors being generated.

For the record, I did that and the output was included in the issue. Nothing was shown for either issue.

**Comment by nhuethmayr** [comment]
> I created this bug even though I find the fix for my case because the lack of accurate error reporting from the emulator cost me tons of time here. And no doubt, other people as well.
> 
> The fix in my case was to change the `tsconfig.json` from
> 
> ```
>     "module": "esnext",
>     "moduleResolution": "node",
> ```
> 
> to
> 
> ```
>     "module": "NodeNext",
>     "moduleResolution": "NodeNext",
> ```
> 
> Is there a technical reason why the emulator couldn't advise me of that?

Changing the **module/moduleResolution** worked well for me.

**Comment by codebravotech** [comment]
For others who get this issue, changing the tsconfig to the above from the settings that are written by the firebase functions initialization script didn't work for me. 

I ended up realizing that I had an outdated version of firebase-tools installed globally, so I removed that and added the @latest version as a dev dependency to the project itself, and now it runs. 

I am using Typescript so I also changed their `exports.function_name` syntax to `export const function_name`. Hope that can help someone avoid similar hairpulling
